prusa/link/web/main.py

```diff
@@ -428,23 +428,42 @@ def api_update(req, env):
     versions"""
     # pylint: disable=unused-argument
     headers = {"Update-Available": "False"}
-    updatable = []
 
     if env == "python":
-        output = check_output([executable, '-m', 'pip', 'list', '-o']).decode()
-        packages_list = output.split('\n')[2:-1]
-        if packages_list:
-            headers["Update-Available"] = "True"
-        for package_ in packages_list:
-            package = package_.split()
-            updatable.append(
-                {
-                    "name": package[0],
-                    "version": package[1],
-                    "available_version": package[2],
-                    "path": None}
-            )
-        return JSONResponse(available_updates=updatable, headers=headers)
+        try:
+            output = check_output(
+                [executable, '-m', 'pip', 'install', '--dry-run', '-U',
+                 'prusalink']).decode()
+
+        # There's a problem with package installation, or it does not exist
+        except CalledProcessError as exception:
+            raise conditions.UnavailablePackage() from exception
+
+        # New version is available to download and possible to install
+        if "Would install" in output:
+            report = output.splitlines()[-1].split()[-1]
+            return JSONResponse(available_update=report, headers=headers)
+
+        # No update available
+        return Response(status_code=state.HTTP_NO_CONTENT)
+
+    if env == "system":
+        return Response(status_code=state.HTTP_NOT_IMPLEMENTED)
+
+    return Response(status_code=state.HTTP_BAD_REQUEST)
+
+
+@app.route("/api/v1/update/<env>", method=state.METHOD_POST)
+@check_api_digest
+def api_update_post(req, env):
+    """Update all packages in given environment"""
+    # pylint: disable=unused-argument
+    if env == "python":
+        process_call = call(
+            "pip install -U --upgrade-strategy eager prusalink", shell=True)
+        if process_call == 0:
+            return Response(status_code=state.HTTP_OK)
+        raise conditions.UnavailablePackage
```

Comment:
It seams like, only UnavailablePackage can happen. Error must be more common, like UpdateFailed, or something like that.